clients/drcachesim/tracer/tracer.cpp

```diff
@@ -99,6 +103,64 @@ DR_DISALLOW_UNSAFE_STATIC
 
 namespace dynamorio {
 namespace drmemtrace {
+namespace {
+#ifdef BUILD_TRACER_WITH_DR_SYSCALL
+#    define SYSCALL_RECORD_BUFFER_SIZE 8192
+
+file_t syscall_record_file;
+int syscall_record_file_offset = 0;
+char syscall_record_buffer[SYSCALL_RECORD_BUFFER_SIZE];
+
+size_t
+write_syscall_record(char *buf, size_t size)
+{
+    size_t byte_written = 0;
+    size_t remaining = size;
+
+    while (remaining + syscall_record_file_offset >= SYSCALL_RECORD_BUFFER_SIZE) {
+        memcpy(&syscall_record_buffer[syscall_record_file_offset], buf,
+               SYSCALL_RECORD_BUFFER_SIZE - syscall_record_file_offset);
+        const ssize_t wrote =
+            write(syscall_record_file, syscall_record_buffer, SYSCALL_RECORD_BUFFER_SIZE);
```

Comment:
For default use, this should use dr_write_file. But this file i/o needs to all be indirected like all i/o in drmemtrace to support use cases that need library-based filesystems.